pddl/logic/functions.py

```diff
@@ -0,0 +1,366 @@
+            self.optimization in {self.MAXIMIZE, self.MINIMIZE},
+            "Optimization metric not recognized.",
+        )
+
+    def __str__(self) -> str:
+        """Get the string representation."""
+        return f"{self.optimization} {self.expression}"
+
+    def __repr__(self) -> str:
+        """Get an unambiguous string representation."""
+        return f"{type(self).__name__}({self.expression}, {self.optimization})"  # type: ignore
+
+    def __eq__(self, other):
+        """Override equal operator."""
+        return (
+            isinstance(other, Metric)
+            and self.expression == other.expression
+            and self.optimization == other.optimization
+        )
+
+    def __hash__(self):
+        """Get the hash of a Metric."""
+        return hash((self.expression, self.optimization))
+
+
+class EqualTo(BinaryFunction, metaclass=MonotoneOp):
+    """Equal to operator."""
+
+    SYMBOL = Symbols.EQUAL
+
+
+class LesserThan(BinaryFunction):
+    """Lesser than operator."""
+
+    SYMBOL = Symbols.LESSER
+
+
+class LesserEqualThan(BinaryFunction):
+    """Lesser or equal than operator."""
+
+    SYMBOL = Symbols.LESSER_EQUAL
+
+
+class GreaterThan(BinaryFunction):
+    """Greater than operator."""
+
+    SYMBOL = Symbols.GREATER
+
+
+class GreaterEqualThan(BinaryFunction):
+    """Greater or equal than operator."""
+
+    SYMBOL = Symbols.GREATER_EQUAL
+
+
+class Assign(BinaryFunction):
+    """Assign operator."""
+
+    SYMBOL = Symbols.ASSIGN
+
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Assign operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class ScaleUp(BinaryFunction):
+    """Scale-Up operator."""
+
+    SYMBOL = Symbols.SCALE_UP
+
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Scale-Up operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class ScaleDown(BinaryFunction):
+    """Scale-Down operator."""
+
+    SYMBOL = Symbols.SCALE_DOWN
+
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Scale-Down operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Increase(BinaryFunction):
+    """Increase operator."""
+
+    SYMBOL = Symbols.INCREASE
+
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Increase operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Decrease(BinaryFunction):
+    """Decrease operator."""
+
+    SYMBOL = Symbols.DECREASE
+
+    def __init__(self, *operands: FunctionExpression):
+        """
+        Initialize the Decrease operator.
+
+        :param operands: the operands.
+        """
+        super().__init__(*operands)
+
+
+class Minus(BinaryFunction, metaclass=MonotoneOp):
```

Comment:
I fear this is not correct. `MonotoneOp` will apply the idempotency simplification, which makes sense for logical `or` and `and`, but not for mathematical operations on numbers.

i.e. `a & a \equiv a`, but `a - a  \not\equiv a`.

You can see where this is implemented at this line: basically, we remove duplicates.
https://github.com/AI-Planning/pddl/blob/9b57554782f8748870aa4c255f451f0b0382a796/pddl/logic/base.py#L264




```suggestion
class Minus(BinaryFunction):
```

Reply:
That's right. I thought it was useful to flatten the operands.

Reply:
Minimum example to reproduce the problem:

```python
from pddl.logic.functions import Minus, NumericValue
a1 = NumericValue(1)
b1 = Minus(a1, a1, a1)  # this should be Minus, but we get NumericValue
a1 == b1 # we get True, but it should be false
```

Reply:
I think it should be as easy as removing the metaclass specification

Reply:
great, I'll remove the metaclass!

Reply:
> That's right. I thought it was useful to flatten the operands.

this comment actually suggests that we should add a class attribute `idempotency: bool`, which by default is `False`, and change the behaviour of `MonotoneOp` depending on what is the value of `idempotency`. We should change the name for a generic binary operation metaclass, e.g. `BinaryOpMetaclass`. By doing so, we can still gain from the usefulness of flattening.

Reply:
#96